### PR TITLE
+emr serverless +pyspark scene det ^lane det o/p update

### PR DESCRIPTION
**Pull request by manojrajpurohit**
*Issue #, if available:*  N/A

*Description of changes:* 
1. Added EMR Serverless application IaC, it creates an EMR Serverless application. 
2. Updated Scene detection logic using  PySpark native functions. The scene detection logic is generic and thus extensible for any class of object in lane.
3. Updated Lane Detection output format, Before: The lane detection output was persisting a tensor which requires tensor libraries to be imported in any downstream analytics application to read tensor objects. Now: The lane detection output format is a nested list of pixel intensity 0 representing background and 1 representing a lane pixel. Further, the lane detection output writes all lane outputs as rows in a csv file, earlier output format of 1 csv per image is not efficient to be read by downstream analytical applications - small file problem.

**Important Note:**  The pyspark script has a dependency on as JAR spark-dynamodb_2.12 download link -> https://repo1.maven.org/maven2/com/audienceproject/spark-dynamodb_2.12/1.1.1/spark-dynamodb_2.12-1.1.1.jar . This Jar should be manually uploaded to following S3 prefix path s3://addf-ros-image-demo-artifacts-bucket-[random characters]/dags/ros-image-demo/dags-aws/spark_scripts/

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: modules/core/mwaa/requirements/requirements.txt [hunk 1]
```diff
@@ -1,8 +1,9 @@
 Pillow~=9.0.1
-boto3~=1.21.37
-awscli~=1.22.80
+boto3~=1.23.9
+awscli~=1.24.9
```
Comment (dgraeber) [inline modules/core/mwaa/requirements/requirements.txt:3]: the `core/mwaa` module now supports passing in a requirements.txt file to configure MWAA.  Please leverage this capability and not modify the existing code as we need to support backward-compatability

**Review by dgraeber — CHANGES_REQUESTED** [review]
I can review this after the holidays, but this is missing:

1. a CHANGELOG
2. a workflow in the git modules (`.github/workflows/main.yml`)
3. the deployspec should be handling the upload of the dependency jar as referenced in your PR description
4. VPC support as this is a core module - how do you also support integration with other services?
5. fix.sh and validate.sh should be run...these are unformatted

This PR shouldn't change existing configurations (ref core/mwaa).  I implemented the ability to designate which requirement.txt to install on MWAA and your PR should leverage it.

See me for details?

File: modules/core/emr-serverless/stack.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        name=dep_mod
+        )
+        
+        policy_statements = [
+            iam.PolicyStatement(
+                actions=["dynamodb:*"],
+                effect=iam.Effect.ALLOW,
+                resources=[f"arn:aws:dynamodb:{self.region}:{self.account}:table/addf*"],
+            ),
+            iam.PolicyStatement(
```
Comment (dgraeber) [inline modules/core/emr-serverless/stack.py:81]: To wide open....

File: modules/core/emr-serverless/stack.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            memory="100GB",
+    
+            # the properties below are optional
+            disk="1000GB"
+        ),
+        name=dep_mod
+        )
+        
+        policy_statements = [
+            iam.PolicyStatement(
```
Comment (dgraeber) [inline modules/core/emr-serverless/stack.py:76]: Too wide open

File: modules/core/emr-serverless/stack.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            enabled=False,
+            idle_timeout_minutes=5
+        ),
+        initial_capacity=
+        [emrserverless.CfnApplication.InitialCapacityConfigKeyValuePairProperty(
+            key="Driver",
+            value=emrserverless.CfnApplication.InitialCapacityConfigProperty(
+                worker_configuration
+                =emrserverless.CfnApplication.WorkerConfigurationProperty(
+                    cpu="2vCPU",
```
Comment (dgraeber) [inline modules/core/emr-serverless/stack.py:40]: Should these be configurable?

File: modules/core/emr-serverless/stack.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                ),
+                worker_count=2
+            )
+        ),
+        emrserverless.CfnApplication.InitialCapacityConfigKeyValuePairProperty(
+            key="Executor",
+            value=emrserverless.CfnApplication.InitialCapacityConfigProperty(
+                worker_configuration
+                =emrserverless.CfnApplication.WorkerConfigurationProperty(
+                    cpu="2vCPU",
```
Comment (dgraeber) [inline modules/core/emr-serverless/stack.py:54]: Configurable?

File: modules/core/emr-serverless/app.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+deployment_name = os.getenv("ADDF_DEPLOYMENT_NAME", "")
+module_name = os.getenv("ADDF_MODULE_NAME", "")
+
+def _param(name: str) -> str:
+    return f"ADDF_PARAMETER_{name}"
+
+app = App()
+
+emr_serverless = EmrServerlessStack(
```
Comment (dgraeber) [inline modules/core/emr-serverless/app.py:15]: This should run in a VPC and leverage VPC endpoints?

File: modules/core/emr-serverless/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+- creates an Amazon EMR Serverless Application
+- creates an IAM Role and policy for Jobs on the EMR Application
+
+## Inputs/Outputs
+
+### Input Paramenters
+
+#### Required
+
+- None
```
Comment (dgraeber) [inline modules/core/emr-serverless/README.md:16]: A core module should support VPC integration.

File: modules/analysis/rosbag-image-pipeline/image_dags/ros_image_pipeline.py [hunk 8]
```diff
@@ -414,15 +441,45 @@ def sagemaker_lanedet_operation(**kwargs):
...
+                "--batch-id", batch_id,
+                "--bucket", TARGET_BUCKET,
+                "--region", REGION,
+                "--output-dynamo-table", SCENE_TABLE
+            ],
+            "sparkSubmitParameters": "--jars s3://addf-ros-image-demo-artifacts-bucket-75fe6115/dags/ros-image-demo/dags-aws/spark_scripts/spark-dynamodb_2.12-1.1.1.jar"
+        }
+    }
+
+    start_job_run_op = EmrServerlessStartJobOperator(
```
Comment (dgraeber) [inline modules/analysis/rosbag-image-pipeline/image_dags/ros_image_pipeline.py:462]: This DAG has existed prior to EMR Serverless support...?  If so, is this change backward-compatible?  Will it break the existing pipeline usage?

**Comment by dgraeber** [comment]
@manojrajpurohit please see above. How would you like to proceed?

**Comment by dgraeber** [comment]
Closing this PR due to inactivity.  Once a design session / KT session can be conducted and the issues remediated, we can reopen.

**Comment by srinivasreddych** [comment]
Reopening this PR...

**Comment by srinivasreddych** [comment]
Closing this in favor of https://github.com/awslabs/idf-modules/pull/29 and eventually splitting the business logic into future PRs

**Pull request closed without merging**